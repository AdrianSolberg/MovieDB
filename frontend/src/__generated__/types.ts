export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type MakeEmpty<T extends { [key: string]: unknown }, K extends keyof T> = { [_ in K]?: never };
export type Incremental<T> = T | { [P in keyof T]?: P extends ' $fragmentName' | '__typename' ? T[P] : never };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: { input: string; output: string; }
  String: { input: string; output: string; }
  Boolean: { input: boolean; output: boolean; }
  Int: { input: number; output: number; }
  Float: { input: number; output: number; }
  Date: { input: any; output: any; }
};

export type Movie = {
  _id: Scalars['Int']['output'];
  title: Scalars['String']['output'];
  vote_average: Scalars['Float']['output'];
  vote_count: Scalars['Int']['output'];
  status: Scalars['String']['output'];
  release_date: Scalars['Date']['output'];
  revenue: Scalars['Float']['output'];
  runtime: Scalars['Int']['output'];
  backdrop_path?: Maybe<Scalars['String']['output']>;
  budget: Scalars['Int']['output'];
  homepage?: Maybe<Scalars['String']['output']>;
  imdb_id?: Maybe<Scalars['String']['output']>;
  original_language: Scalars['String']['output'];
  original_title: Scalars['String']['output'];
  overview: Scalars['String']['output'];
  popularity: Scalars['Float']['output'];
  poster_path?: Maybe<Scalars['String']['output']>;
  tagline?: Maybe<Scalars['String']['output']>;
  genres: Array<Scalars['String']['output']>;
  production_companies: Array<Scalars['String']['output']>;
  production_countries: Array<Scalars['String']['output']>;
  spoken_languages: Array<Scalars['String']['output']>;
  keywords: Array<Scalars['String']['output']>;
};

export type Filters = {
  Genre: Array<Scalars['String']['output']>;
  Rating: Array<Scalars['String']['output']>;
  Decade: Array<Scalars['String']['output']>;
  Status: Array<Scalars['String']['output']>;
  Runtime: Array<Scalars['String']['output']>;
};

export type FiltersInput = {
  Genre: Array<Scalars['String']['input']>;
  Rating: Array<Scalars['String']['input']>;
  Decade: Array<Scalars['String']['input']>;
  Status: Array<Scalars['String']['input']>;
  Runtime: Array<Scalars['String']['input']>;
};

export enum SortingType {
  NewestFirst = 'NEWEST_FIRST',
  OldestFirst = 'OLDEST_FIRST',
  BestRated = 'BEST_RATED',
  WorstRated = 'WORST_RATED',
  LongestRuntime = 'LONGEST_RUNTIME',
  ShortestRuntime = 'SHORTEST_RUNTIME'
}

export type Query = {
  movie?: Maybe<Movie>;
  movies: Array<Movie>;
  filters?: Maybe<Filters>;
};


export type QueryMovieArgs = {
  id: Scalars['Int']['input'];
};


export type QueryMoviesArgs = {
  skip?: InputMaybe<Scalars['Int']['input']>;
  limit?: InputMaybe<Scalars['Int']['input']>;
  filters?: InputMaybe<FiltersInput>;
  sortOption?: InputMaybe<SortingType>;
  search?: InputMaybe<Scalars['String']['input']>;
};

export type GetMovieQueryVariables = Exact<{
  movieId: Scalars['Int']['input'];
}>;


export type GetMovieQuery = { __typename?: 'Query', movie?: { __typename?: 'Movie', _id: number, title: string, vote_average: number, vote_count: number, status: string, release_date: any, revenue: number, runtime: number, backdrop_path?: string | null, budget: number, homepage?: string | null, imdb_id?: string | null, original_language: string, original_title: string, overview: string, popularity: number, poster_path?: string | null, tagline?: string | null, genres: Array<string>, production_companies: Array<string>, production_countries: Array<string>, spoken_languages: Array<string>, keywords: Array<string> } | null };

export type GetMoviesQueryVariables = Exact<{
  skip?: InputMaybe<Scalars['Int']['input']>;
  limit?: InputMaybe<Scalars['Int']['input']>;
  filters?: InputMaybe<FiltersInput>;
  sortOption?: InputMaybe<SortingType>;
  search?: InputMaybe<Scalars['String']['input']>;
}>;


export type GetMoviesQuery = { __typename?: 'Query', movies: Array<{ __typename?: 'Movie', _id: number, title: string, vote_average: number, release_date: any, runtime: number, poster_path?: string | null }> };

export type GetFiltersQueryVariables = Exact<{ [key: string]: never; }>;


export type GetFiltersQuery = { __typename?: 'Query', filters?: { __typename?: 'Filters', Genre: Array<string>, Rating: Array<string>, Decade: Array<string>, Status: Array<string>, Runtime: Array<string> } | null };
