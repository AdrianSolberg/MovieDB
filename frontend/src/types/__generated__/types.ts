export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = {
    [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & {
    [SubKey in K]?: Maybe<T[SubKey]>;
};
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & {
    [SubKey in K]: Maybe<T[SubKey]>;
};
export type MakeEmpty<
    T extends { [key: string]: unknown },
    K extends keyof T,
> = { [_ in K]?: never };
export type Incremental<T> =
    | T
    | {
          [P in keyof T]?: P extends " $fragmentName" | "__typename"
              ? T[P]
              : never;
      };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
    ID: { input: string; output: string };
    String: { input: string; output: string };
    Boolean: { input: boolean; output: boolean };
    Int: { input: number; output: number };
    Float: { input: number; output: number };
    DateTime: { input: Date; output: Date };
};

export type Movie = {
    __typename?: "Movie";
    _id: Scalars["Int"]["output"];
    title: Scalars["String"]["output"];
    vote_average: Scalars["Float"]["output"];
    vote_count: Scalars["Int"]["output"];
    status: Scalars["String"]["output"];
    release_date: Scalars["DateTime"]["output"];
    revenue: Scalars["Float"]["output"];
    runtime: Scalars["Int"]["output"];
    backdrop_path?: Maybe<Scalars["String"]["output"]>;
    budget: Scalars["Int"]["output"];
    homepage?: Maybe<Scalars["String"]["output"]>;
    imdb_id?: Maybe<Scalars["String"]["output"]>;
    original_language: Scalars["String"]["output"];
    original_title: Scalars["String"]["output"];
    overview: Scalars["String"]["output"];
    popularity: Scalars["Float"]["output"];
    poster_path?: Maybe<Scalars["String"]["output"]>;
    tagline?: Maybe<Scalars["String"]["output"]>;
    genres: Array<Scalars["String"]["output"]>;
    production_companies: Array<Scalars["String"]["output"]>;
    production_countries: Array<Scalars["String"]["output"]>;
    spoken_languages: Array<Scalars["String"]["output"]>;
    keywords: Array<Scalars["String"]["output"]>;
    reviews: Array<Review>;
};

export type Review = {
    __typename?: "Review";
    _id: Scalars["ID"]["output"];
    movie: Movie;
    username: Scalars["String"]["output"];
    rating: Scalars["Int"]["output"];
    comment: Scalars["String"]["output"];
    date: Scalars["DateTime"]["output"];
};

export type Filters = {
    __typename?: "Filters";
    Genre: Array<Scalars["String"]["output"]>;
    Rating: Array<Scalars["String"]["output"]>;
    Decade: Array<Scalars["String"]["output"]>;
    Status: Array<Scalars["String"]["output"]>;
    Runtime: Array<Scalars["String"]["output"]>;
};

export type FiltersInput = {
    Genre: Array<Scalars["String"]["input"]>;
    Rating: Array<Scalars["String"]["input"]>;
    Decade: Array<Scalars["String"]["input"]>;
    Status: Array<Scalars["String"]["input"]>;
    Runtime: Array<Scalars["String"]["input"]>;
};

export enum SortingType {
    MOST_POPULAR = "MOST_POPULAR",
    NEWEST_FIRST = "NEWEST_FIRST",
    OLDEST_FIRST = "OLDEST_FIRST",
    BEST_RATED = "BEST_RATED",
    WORST_RATED = "WORST_RATED",
    LONGEST_RUNTIME = "LONGEST_RUNTIME",
    SHORTEST_RUNTIME = "SHORTEST_RUNTIME",
}

export type Query = {
    __typename?: "Query";
    movie?: Maybe<Movie>;
    movies: Array<Movie>;
    filters: Filters;
    latestReviews: Array<Review>;
    userReviews: Array<Review>;
};

export type QuerymovieArgs = {
    id: Scalars["Int"]["input"];
};

export type QuerymoviesArgs = {
    skip?: InputMaybe<Scalars["Int"]["input"]>;
    limit?: InputMaybe<Scalars["Int"]["input"]>;
    filters?: InputMaybe<FiltersInput>;
    sortOption?: InputMaybe<SortingType>;
    search?: InputMaybe<Scalars["String"]["input"]>;
};

export type QuerylatestReviewsArgs = {
    skip?: InputMaybe<Scalars["Int"]["input"]>;
    limit?: InputMaybe<Scalars["Int"]["input"]>;
};

export type QueryuserReviewsArgs = {
    username: Scalars["String"]["input"];
    skip?: InputMaybe<Scalars["Int"]["input"]>;
    limit?: InputMaybe<Scalars["Int"]["input"]>;
};

export type Mutation = {
    __typename?: "Mutation";
    addReview: Movie;
    deleteReview: Movie;
};

export type MutationaddReviewArgs = {
    movie_id: Scalars["Int"]["input"];
    username: Scalars["String"]["input"];
    rating: Scalars["Int"]["input"];
    comment: Scalars["String"]["input"];
};

export type MutationdeleteReviewArgs = {
    _id: Scalars["ID"]["input"];
};

export type GetMoviesQueryVariables = Exact<{
    skip?: InputMaybe<Scalars["Int"]["input"]>;
    limit?: InputMaybe<Scalars["Int"]["input"]>;
    filters?: InputMaybe<FiltersInput>;
    sortOption?: InputMaybe<SortingType>;
    search?: InputMaybe<Scalars["String"]["input"]>;
}>;

export type GetMoviesQuery = {
    __typename?: "Query";
    movies: Array<{
        __typename?: "Movie";
        _id: number;
        title: string;
        vote_average: number;
        release_date: Date;
        runtime: number;
        poster_path?: string | null;
    }>;
};

export type GetMovieQueryVariables = Exact<{
    movieId: Scalars["Int"]["input"];
}>;

export type GetMovieQuery = {
    __typename?: "Query";
    movie?: {
        __typename?: "Movie";
        _id: number;
        title: string;
        vote_average: number;
        vote_count: number;
        status: string;
        release_date: Date;
        revenue: number;
        runtime: number;
        backdrop_path?: string | null;
        budget: number;
        homepage?: string | null;
        imdb_id?: string | null;
        original_language: string;
        original_title: string;
        overview: string;
        popularity: number;
        poster_path?: string | null;
        tagline?: string | null;
        genres: Array<string>;
        production_companies: Array<string>;
        production_countries: Array<string>;
        spoken_languages: Array<string>;
        keywords: Array<string>;
        reviews: Array<{
            __typename?: "Review";
            _id: string;
            username: string;
            rating: number;
            comment: string;
            date: Date;
        }>;
    } | null;
};

export type GetFiltersQueryVariables = Exact<{ [key: string]: never }>;

export type GetFiltersQuery = {
    __typename?: "Query";
    filters: {
        __typename?: "Filters";
        Genre: Array<string>;
        Rating: Array<string>;
        Decade: Array<string>;
        Status: Array<string>;
        Runtime: Array<string>;
    };
};

export type AddReviewMutationVariables = Exact<{
    movieId: Scalars["Int"]["input"];
    username: Scalars["String"]["input"];
    rating: Scalars["Int"]["input"];
    comment: Scalars["String"]["input"];
}>;

export type AddReviewMutation = {
    __typename?: "Mutation";
    addReview: {
        __typename?: "Movie";
        reviews: Array<{
            __typename?: "Review";
            _id: string;
            username: string;
            rating: number;
            comment: string;
            date: Date;
        }>;
    };
};
