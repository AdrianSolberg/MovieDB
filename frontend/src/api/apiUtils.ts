import { Review } from "@/types/__generated__/types";

type ResponseReview = {
    __typename?: "Review";
    _id: string;
    username: string;
    rating: number;
    comment: string;
    date: Date;
};

/**
 *  Make sure that the date-field is correctly initialized as a Date-object
 *  when getting reviews as a response from a query.
 */
export function parseReviewResponse(inputReviews: ResponseReview[]): Review[] {
    return inputReviews.map((review) => ({
        ...review,
        date: new Date(review.date),
    })) as Review[];
}
