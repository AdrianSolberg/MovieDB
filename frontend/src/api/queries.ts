import { gql } from "@/types/__generated__";

export const GET_MOVIES = gql(`
    query GetMovies(
        $skip: Int
        $limit: Int
        $filters: FiltersInput
        $sortOption: SortingType
        $search: String
    ) {
        movies(
            skip: $skip
            limit: $limit
            filters: $filters
            sortOption: $sortOption
            search: $search
        ) {
            _id
            title
            vote_average
            release_date
            runtime
            poster_path
        }
    }
`);

export const GET_MOVIE = gql(`
    query GetMovie($movieId: Int!) {
        movie(id: $movieId) {
            _id
            title
            vote_average
            vote_count
            status
            release_date
            revenue
            runtime
            backdrop_path
            budget
            homepage
            imdb_id
            original_language
            original_title
            overview
            popularity
            poster_path
            tagline
            genres
            production_companies
            production_countries
            spoken_languages
            keywords
            reviews {
                _id
                username
                rating
                comment
                date
            }
        }
    }
`);

export const GET_FILTERS = gql(`
    query GetFilters {
        filters {
            Genre
            Rating
            Decade
            Status
            Runtime
        }
    }
`);

export const GET_LATEST_REVIEWS = gql(`
    query GetLatestReviews(
        $skip: Int,
        $limit: Int
    ) {
        latestReviews(
            skip: $skip,
            limit: $limit
        ) {
            _id
            movie {
                _id
                title
                poster_path
            }
            username
            rating
            comment
            date
        }
    }
`);

export const GET_USER_REVIEWS = gql(`
    query GetUserReviews(
        $username: String!,
        $skip: Int,
        $limit: Int
    ) {
        userReviews(
            username: $username,
            skip: $skip,
            limit: $limit
        ) {
            _id
            movie {
                _id
                title
                poster_path
            }
            rating
            comment
            date
        }
    }
`);

export const ADD_REVIEW = gql(`
    mutation AddReview(
        $movieId: Int!,
        $username: String!,
        $rating: Int!,
        $comment: String!
    ) {
        addReview(
            movie_id: $movieId,
            username: $username,
            rating: $rating,
            comment: $comment
        ) {
            _id
            vote_average
            vote_count
            reviews {
                _id
                username
                rating
                comment
                date
            }
        }
    }
`);

export const DELETE_REVIEW = gql(`
    mutation DeleteReview($id: ID!) {
        deleteReview(_id: $id) {
            _id
            vote_average
            vote_count
            reviews {
                _id
                username
                rating
                comment
                date
            }
        }
    }
`);
