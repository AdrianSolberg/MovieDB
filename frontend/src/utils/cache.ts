import { makeVar } from "@apollo/client";
import { Filters, SortingType } from "@/types/movieTypes";

export const sortOptionVar = makeVar<SortingType>(
    (sessionStorage.getItem("sort_option") as SortingType) ??
        SortingType.NEWEST_FIRST
);


const storedFilters = sessionStorage.getItem("filters");
const initialFilters: Filters = storedFilters ? JSON.parse(storedFilters) : {
    Genre: [],
    Rating: [],
    Decade: [],
    Status: [],
    Runtime: [],
};
export const filtersVar = makeVar<Filters>(initialFilters);

export const searchVar = makeVar<string>(
    sessionStorage.getItem("search") ?? ""
);

export const usernameVar = makeVar<string>(
    localStorage.getItem("username") ?? "Guest"
);
