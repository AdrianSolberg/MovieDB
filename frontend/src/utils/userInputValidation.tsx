import { toast } from "sonner";

/**
 *  Returns true if the username is valid, false otherwise.
 *  A valid username must be between 3 and 20 characters long.
 *
 *  Displays a toast error if the username is too short or too long.
 */
export function validateUsername(username: string) {
    const usernameMinLength = 3;
    const usernameMaxLength = 20;

    if (username.length < usernameMinLength) {
        toast.error("Username is too short", {
            description: `Username must be at least ${usernameMinLength} characters long`,
        });
        return false;
    }

    if (username.length > usernameMaxLength) {
        toast.error("Username is too long", {
            description: `Username must be at most  ${usernameMaxLength} characters long`,
        });
        return false;
    }

    return true;
}

/**
 * Returns true if the review is valid, false otherwise.
 * A valid review must be at most 1500 characters long.
 *
 * Displays a toast error if the review is too long.
 */
export function validateReview(comment: string) {
    const reviewMaxLength = 1500;

    if (comment.trim().length > reviewMaxLength) {
        toast.error("Review is too long", {
            description: `Review must be at most ${reviewMaxLength} characters.`,
        });
        return false;
    }

    return true;
}
