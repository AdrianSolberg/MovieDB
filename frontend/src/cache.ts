import { makeVar } from "@apollo/client";
import { SortingType } from "./utils/searchSortAndFilter";

export const sortOptionVar = makeVar<SortingType>(
    (sessionStorage.getItem("sort_option") as SortingType) ??
        SortingType.NEWEST_FIRST
);

export const filtersVar = makeVar<{ [key: string]: string[] }>(
    JSON.parse(sessionStorage.getItem("filters") ?? "{}")
);

export const searchVar = makeVar<string>(
    sessionStorage.getItem("search") ?? ""
);
