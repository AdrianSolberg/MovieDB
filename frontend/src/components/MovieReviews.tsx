import React, { useState, useEffect } from "react";
import { Button } from "../shadcn/components/ui/button";
import { Textarea } from "../shadcn/components/ui/textarea";
import { Card, CardContent } from "../shadcn/components/ui/card";
import Ratings from "../shadcn/components/ui/rating";
import { usernameVar } from "@/utils/cache";
import { useReactiveVar } from "@apollo/client";
import { Movie, Review } from "@/types/__generated__/types";
interface MovieReviewsProps {
    movie: Movie;
}

const MovieReviews: React.FC<MovieReviewsProps> = ({ movie }) => {
    const [reviews, setReviews] = useState<Review[]>([]);
    const [rating, setRating] = useState<number>(0);
    const [comment, setComment] = useState<string>("");
    const username = useReactiveVar(usernameVar);

    useEffect(() => {
        const storedReviews = localStorage.getItem(`movieReviews_${movie._id}`);
        if (storedReviews) {
            const reviews = JSON.parse(storedReviews) as Array<
                Omit<Review, "date" | "movie"> & {
                    date: string;
                    movie: Omit<Movie, "release_date"> & {
                        release_date: string;
                    };
                }
            >;
            setReviews(
                reviews.map((review) => ({
                    ...review,
                    date: new Date(review.date),
                    movie: {
                        ...review.movie,
                        release_date: new Date(review.movie.release_date),
                    },
                }))
            );
        }
    }, [movie]);

    const saveReviews = (updatedReviews: Review[]) => {
        localStorage.setItem(
            `movieReviews_${movie._id}`,
            JSON.stringify(updatedReviews)
        );
        setReviews(updatedReviews);
    };

    const handleSubmitReview = (e: React.FormEvent) => {
        e.preventDefault();
        const newReview: Review = {
            _id: Date.now().toString(), //Temporary id
            movie: movie,
            username: username,
            rating,
            comment: comment.trim(),
            date: new Date(),
        };
        const updatedReviews = [...reviews, newReview];
        saveReviews(updatedReviews);
        setRating(0);
        setComment("");
    };

    const handleDeleteReview = (review: Review) => {
        const updatedReviews = reviews.filter((r) => r._id !== review._id);
        saveReviews(updatedReviews);
    };

    const formatDate = (date: Date) => {
        return date.toLocaleDateString("en-US", {
            year: "numeric",
            month: "long",
            day: "numeric",
            hour: "2-digit",
            minute: "2-digit",
        });
    };

    return (
        <Card className="m-4">
            <section className="m-4 mb-6">
                <h3 className="mb-4 text-2xl font-bold">Submit review</h3>
                <form onSubmit={handleSubmitReview} className="space-y-4">
                    <Ratings
                        value={rating}
                        onValueChange={setRating}
                        variant="yellow"
                        totalstars={5}
                        size={24}
                        asInput={true}
                    />
                    <Textarea
                        id="review-comment"
                        name="comment"
                        value={comment}
                        onChange={(e) => setComment(e.target.value)}
                        placeholder="Write your review (optional)"
                    />
                    <Button type="submit" disabled={rating === 0}>
                        Submit Review
                    </Button>
                </form>
            </section>
            <section className="m-4 space-y-4">
                {reviews.length > 0 && (
                    <h3 className="mb-4 text-2xl font-bold">Reviews</h3>
                )}
                {reviews.map((review) => (
                    <Card key={review._id}>
                        <CardContent className="mt-4">
                            <section className="mb-2 flex items-start justify-between">
                                <section>
                                    <h4 className="text-xl font-bold">
                                        {review.username}
                                    </h4>
                                    <time className="text-sm text-gray-500">
                                        {formatDate(review.date)}
                                    </time>
                                </section>
                                {review.username !== "Guest" &&
                                    username === review.username && (
                                        <Button
                                            onClick={() =>
                                                handleDeleteReview(review)
                                            }
                                        >
                                            Delete
                                        </Button>
                                    )}
                            </section>
                            <Ratings
                                value={review.rating}
                                variant="yellow"
                                totalstars={5}
                            />
                            {review.comment !== "" && (
                                <p className="mt-2">{review.comment}</p>
                            )}
                        </CardContent>
                    </Card>
                ))}
            </section>
        </Card>
    );
};

export default MovieReviews;
