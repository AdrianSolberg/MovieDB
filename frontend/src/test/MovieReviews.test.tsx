import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { MockedProvider } from "@apollo/client/testing";
import userEvent from "@testing-library/user-event";
import MovieReviews from "../components/MovieReviews";
import { ADD_REVIEW } from "@/api/queries";
import { Review } from "@/types/__generated__/types";
import { usernameVar } from "@/utils/cache";
import { all_movies } from "./mock/util";
import { vi } from "vitest";
import { useReactiveVar } from "@apollo/client";
import { validateReview, validateUsername } from "@/utils/userInputValidation";

vi.mock("../components/ReviewCard", () => ({
    default: vi.fn(({ review }: { review: Review }) => (
        <div data-testid="review-card">
            <span data-testid="review-username">{review.username}</span>
            <span data-testid="review-rating">{review.rating}</span>
            <span data-testid="review-comment">{review.comment}</span>
        </div>
    )),
}));

vi.mock("../components/Loader", () => ({
    default: vi.fn(() => <div data-testid="loader" />),
}));

vi.mock("@/utils/userInputValidation", () => ({
    validateReview: vi.fn(),
    validateUsername: vi.fn(),
}));

vi.mock("@/utils/cache", () => ({
    usernameVar: vi.fn(),
}));

vi.mock("@apollo/client", async () => {
    const original = await vi.importActual("@apollo/client");
    return {
        ...original,
        useReactiveVar: vi.fn(),
    };
});

const mockDate = new Date("2024-01-01T12:00:00Z");

const mockMovie = {
    ...all_movies[18],
    reviews: [
        {
            _id: "1",
            username: "testuser",
            rating: 4,
            comment: "Great movie!",
            date: mockDate,
        },
    ] as Review[],
};

const mockAddReviewMutation = {
    request: {
        query: ADD_REVIEW,
        variables: {
            movieId: 475557,
            username: "testuser",
            rating: 5,
            comment: "Excellent movie!",
        },
    },
    result: {
        data: {
            addReview: {
                _id: "2",
                username: "testuser",
                rating: 5,
                comment: "Excellent movie!",
                date: mockDate,
                movie: {
                    _id: 475557,
                    vote_average: 4.5,
                    vote_count: 2,
                },
            },
        },
    },
};

const mockNullAddReviewMutation = {
    request: {
        query: ADD_REVIEW,
        variables: {
            movieId: 475557,
            username: "testuser",
            rating: 5,
            comment: "Excellent movie!",
        },
    },
    result: {
        data: {
            addReview: null,
        },
    },
};

describe("MovieReviews", () => {
    beforeAll(() => {
        vi.mocked(validateUsername).mockReturnValue(true);
        vi.mocked(validateReview).mockReturnValue(true);
        vi.mocked(usernameVar).mockReturnValue("testuser");
        vi.mocked(useReactiveVar).mockImplementation((varFn) => varFn());
    });

    afterAll(() => {
        vi.clearAllMocks();
    });

    it("matches snapshot", () => {
        const { asFragment } = render(
            <MockedProvider mocks={[]} addTypename={false}>
                <MovieReviews movie={mockMovie} />
            </MockedProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders existing reviews", async () => {
        render(
            <MockedProvider mocks={[]} addTypename={false}>
                <MovieReviews movie={mockMovie} />
            </MockedProvider>
        );

        expect(screen.getByText("Reviews")).toBeInTheDocument();
        expect(screen.getByText("Great movie!")).toBeInTheDocument();
        expect(screen.getByText("testuser")).toBeInTheDocument();
    });

    it("displays review submission form", async () => {
        render(
            <MockedProvider mocks={[]} addTypename={false}>
                <MovieReviews movie={mockMovie} />
            </MockedProvider>
        );

        expect(screen.getByText("Submit review")).toBeInTheDocument();
        expect(
            screen.getByRole("button", { name: "Submit Review" })
        ).toBeDisabled();
        expect(
            screen.getByPlaceholderText("Write your review (optional)")
        ).toBeInTheDocument();
    });

    it("enables submit button when rating is selected", async () => {
        render(
            <MockedProvider mocks={[]} addTypename={false}>
                <MovieReviews movie={mockMovie} />
            </MockedProvider>
        );

        const ratingStars = screen.getAllByRole("radio");
        await userEvent.click(ratingStars[4]);

        expect(
            screen.getByRole("button", { name: "Submit Review" })
        ).toBeEnabled();
    });

    it("resets review submission form on successful submission", async () => {
        render(
            <MockedProvider mocks={[mockAddReviewMutation]} addTypename={false}>
                <MovieReviews movie={mockMovie} />
            </MockedProvider>
        );

        const ratingStars = screen.getAllByRole("radio");
        await userEvent.click(ratingStars[4]);
        await userEvent.type(
            screen.getByPlaceholderText("Write your review (optional)"),
            "Excellent movie!"
        );
        await userEvent.click(
            screen.getByRole("button", { name: "Submit Review" })
        );

        expect(
            screen.getByPlaceholderText("Write your review (optional)")
        ).toHaveValue("");
    });

    it("does not reset review submission form on null response", async () => {
        render(
            <MockedProvider
                mocks={[mockNullAddReviewMutation]}
                addTypename={false}
            >
                <MovieReviews movie={mockMovie} />
            </MockedProvider>
        );

        const ratingStars = screen.getAllByRole("radio");
        await userEvent.click(ratingStars[4]);
        await userEvent.type(
            screen.getByPlaceholderText("Write your review (optional)"),
            "Excellent movie!"
        );
        await userEvent.click(
            screen.getByRole("button", { name: "Submit Review" })
        );

        expect(
            screen.getByPlaceholderText("Write your review (optional)")
        ).toHaveValue("Excellent movie!");
    });
});
