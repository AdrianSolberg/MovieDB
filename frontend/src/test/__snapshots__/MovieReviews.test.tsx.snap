// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MovieReviews > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="rounded-lg border bg-card text-card-foreground shadow-sm m-4"
  >
    <section
      class="m-4 mb-6"
    >
      <h2
        class="mb-4 text-2xl font-bold"
      >
        Submit review
      </h2>
      <form
        class="space-y-4"
      >
        <div
          class="flex items-center gap-2"
          role="radiogroup"
        >
          <svg
            aria-checked="false"
            aria-label="1 star"
            class="lucide lucide-star text-yellow-200 cursor-pointer"
            fill="none"
            height="24"
            role="radio"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            tabindex="0"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polygon
              points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
            />
          </svg>
          <svg
            aria-checked="false"
            aria-label="2 star"
            class="lucide lucide-star text-yellow-200 cursor-pointer"
            fill="none"
            height="24"
            role="radio"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            tabindex="0"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polygon
              points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
            />
          </svg>
          <svg
            aria-checked="false"
            aria-label="3 star"
            class="lucide lucide-star text-yellow-200 cursor-pointer"
            fill="none"
            height="24"
            role="radio"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            tabindex="0"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polygon
              points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
            />
          </svg>
          <svg
            aria-checked="false"
            aria-label="4 star"
            class="lucide lucide-star text-yellow-200 cursor-pointer"
            fill="none"
            height="24"
            role="radio"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            tabindex="0"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polygon
              points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
            />
          </svg>
          <svg
            aria-checked="false"
            aria-label="5 star"
            class="lucide lucide-star text-yellow-200 cursor-pointer"
            fill="none"
            height="24"
            role="radio"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            tabindex="0"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polygon
              points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
            />
          </svg>
        </div>
        <textarea
          class="flex min-h-[80px] w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
          id="review-comment"
          maxlength="1500"
          name="comment"
          placeholder="Write your review (optional)"
        />
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
          disabled=""
          type="submit"
        >
          Submit Review
        </button>
      </form>
    </section>
    <section
      class="m-4 space-y-4"
    >
      <h2
        class="mb-4 text-2xl font-bold"
      >
        Reviews
      </h2>
      <ul
        class="space-y-6"
      >
        <li>
          <div
            data-testid="review-card"
          >
            <span
              data-testid="review-username"
            >
              testuser
            </span>
            <span
              data-testid="review-rating"
            >
              4
            </span>
            <span
              data-testid="review-comment"
            >
              Great movie!
            </span>
          </div>
        </li>
      </ul>
    </section>
  </div>
</DocumentFragment>
`;
